### PR TITLE
@kanaabe => Redux QA

### PR DESCRIPTION
**Pull request by eessex**
Fixes the following bugs:
- shows 'add new section' text and section tool if no sections
- z-index issues in autocomplete component
- huge header over hero section on classic articles
- captions are updated on drag/drop in image collection
- can drag/drop article sections

### CODE REVIEW

File: client/components/drag_drop/index.coffee [hunk 1]
```diff
@@ -74,7 +74,7 @@ module.exports = React.createClass
 
           DragTarget {
             key: i + '-' + child.type.displayName + '-target'
-            i: i
+            i: child.props.index or i
             setDragTarget: @setDragTarget
```
Comment (eessex) [inline client/components/drag_drop/index.coffee]: Sometimes drag/drop accepts an array of object of mixed types -- i.e., section list included non-draggable section tools in the same array.  This makes sure that it uses the provided index if given, and ignore non-draggable items when returning the new order.

File: client/apps/edit/components/content/sections/text/index.jsx [hunk 2]
```diff
@@ -100,21 +87,52 @@ export class SectionText extends Component {
...
+    const bodyHasChanged = section.body !== prevProps.section.body && section.body.length > 0
+    const bodyWasSwapped = bodyHasChanged && section.body !== html
+    const startedEditing = editing && editing !== prevProps.editing
+    const stoppedEditing = !editing && editing !== prevProps.editing
+
+    if (startedEditing || bodyWasSwapped) {
+      if (bodyHasChanged) {
+        // Re-initialize editor with new text, happens during
+        // drag/drop or changing edit section w/ handleTab or splitSection
+        this.setEditorStateFromProps()
```
Comment (eessex) [inline client/apps/edit/components/content/sections/text/index.jsx]: Draft works from its own state, and will not reset with new data automatically when the redux props change. This checks if props and state have gotten out of sync, for cases when sections are inserted in the middle of the list, or the order is changed by user interactions.

File: client/apps/edit/components/content/sections/text/test/index.test.js [hunk 3]
```diff
@@ -167,6 +167,68 @@ describe('SectionText', () => {
     })
   })
 
+  // maybeResetEditor
```
Comment (kanaabe) [inline client/apps/edit/components/content/sections/text/test/index.test.js]: Stray?

File: client/components/rich_text/components/paragraph.coffee [hunk 4]
```diff
@@ -60,6 +60,11 @@ module.exports = React.createClass
           new CompositeDecorator Config.decorators(@hasLinks())
         )
 
+  componentDidUpdate: (prevProps) ->
+    if @props.html != prevProps.html && @props.html != @state.html
+      # re-initialize on drag/drop
+      @componentDidMount()
```
Comment (kanaabe) [inline client/components/rich_text/components/paragraph.coffee]: I don't _think_ there's an issue with calling directly with this, but maybe it'd feel more natural to move the work in `componentDidMount` to a `setParagraphState` and call the method. Not a blocker, feel free to follow up on it if you wish.

**Pull request merged**